### PR TITLE
ENH: Pluggable SQL performance via new SQL `engine` keyword

### PR DESCRIPTION
**Pull request by yehoshuadimarsky**
- [x] xref, but is first step in implementing #36893
- [x] tests added and passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry (in the docstring)
- [ ] whatsnew entry (in the docs)

This is the first step in implementing an `engine` backend to the SQL I/O functions, as discussed in #36893. This PR is simply to refactor the existing SQLAlchemy parts to make them extensible for an external engine.

No new engines were added at this stage. No new tests were added, only verified that existing tests fail ("test behavior, not implementation" suggests that we don't need to test internal refactoring)

Lots of the code was borrowed from the Parquet I/O implementation.

### CODE REVIEW

**Comment by yehoshuadimarsky** [comment]
weird, looks like CI only has one error but it's totally unrelated - something about IPython code blocks in the docs, but it's in the Parquet section, nothing to do with this PR...

File: pandas/io/sql.py
Comment (jreback) [inline pandas/io/sql.py:704]: -> 1.3

File: pandas/io/sql.py
Comment (jreback) [inline pandas/io/sql.py:2162]: can we avoid kwargs ?

File: pandas/io/sql.py
Comment (jreback) [inline pandas/io/sql.py:1748]: instead use `engine_kwargs`  which we use elsewhere

File: pandas/io/sql.py
Comment (jreback) [inline pandas/io/sql.py:1742]: i think you need a blank line after the version added

File: pandas/io/sql.py
Comment (jreback) [inline pandas/io/sql.py:1383]: can you change for now

Comment (jreback) [inline pandas/io/sql.py:1383]: do we have a test for this? (e.g. engine not one of sqlalchemy / auto)

File: pandas/io/sql.py
Comment (jreback) [inline pandas/io/sql.py:1380]: can be eliif

**Review by jreback — CHANGES_REQUESTED** [review]
looks reasonable. some changes, would be ok with merging this. then if you could refactor pandas/io/sqlpy to make a sub-module e.g. 
pandas/io/sql/engine, .....etc (similar in concept to how we do pandas/io/excel).

could also do that as a pre-cursor here.

**Comment by jreback** [comment]
@pandas-dev/pandas-core if any comments on the structure.

File: pandas/core/config_init.py
Comment (jreback) [inline pandas/core/config_init.py:663]: can you add this to options.rst as well

Comment (yehoshuadimarsky) [reply pandas/core/config_init.py:663]: fixed

File: pandas/io/sql.py
Comment (yehoshuadimarsky) [reply pandas/io/sql.py]: fixed

File: pandas/io/sql.py
Comment (yehoshuadimarsky) [reply pandas/io/sql.py]: yup

File: pandas/io/sql.py
Comment (yehoshuadimarsky) [reply pandas/io/sql.py]: yup

File: pandas/io/sql.py
Comment (yehoshuadimarsky) [reply pandas/io/sql.py:1745]: yup

File: pandas/io/sql.py
Comment (yehoshuadimarsky) [reply pandas/io/sql.py]: ok

File: pandas/io/sql.py
Comment (yehoshuadimarsky) [reply pandas/io/sql.py:2166]: ok

File: pandas/io/sql.py
Comment (yehoshuadimarsky) [reply pandas/io/sql.py]: I tried now, copied mostly from the Parquet tests, but can't figure out where the existing tests use a SQLAlchemy connection that I can reuse. I need it to be SQLAlchemy because otherwise the option io.sql.engine isn't checked.

**Comment by yehoshuadimarsky** [comment]
Made all requested changes except the test, as noted above

File: pandas/io/sql.py
Comment (charlesdong1991) [inline pandas/io/sql.py:1359]: can this be put to the if-clause below together with codes for `engine=='auto'`?

File: pandas/io/sql.py
Comment (charlesdong1991) [inline pandas/io/sql.py:1370]: I am just curious why put it to a list and iterate this one element? what's the difference between this and simply using `SQLAlchemyEngine()` in try-except clause?

Comment (yehoshuadimarsky) [reply pandas/io/sql.py:1370]: Because very soon we will add other engines. I adopted this code from our Parquet file where they do the same thing
https://github.com/pandas-dev/pandas/blob/2019e7b71f392a2790d95fa26b11fe917946e1d1/pandas/io/parquet.py#L41-L72

File: pandas/io/sql.py
Comment (yehoshuadimarsky) [reply pandas/io/sql.py:1359]: Copied from Parquet implementation, see comment below. I assume it was done this way in case the `option` is set to `auto`

File: pandas/io/sql.py
Comment (charlesdong1991) [reply pandas/io/sql.py:1370]: ah, I see! Thanks for your reply, I missed whole picture

File: pandas/io/sql.py
Comment (yehoshuadimarsky) [reply pandas/io/sql.py]: just committed some test changes, pls let me know what you think. Copied/pasted and modified from the Parquet engine tests

**Comment by yehoshuadimarsky** [comment]
Weird CI failure, unrelated to my changes - something to do with `inf` in MySQL - but I think it should be good to merge now, can you pls take a look? Thanks. @jreback

File: pandas/io/sql.py
Comment (jreback) [inline pandas/io/sql.py:1750]: 1.3

File: pandas/io/sql.py
Comment (jreback) [reply pandas/io/sql.py:2166]: can you simply pass engine here

File: pandas/tests/io/test_sql.py
Comment (jreback) [inline pandas/tests/io/test_sql.py:2108]: can you add a test here which raises if the io.sql.engine is specified but not supported

**Review by jreback — CHANGES_REQUESTED** [review]
couple of comments, pls merge master.

File: pandas/tests/io/test_sql.py
Comment (yehoshuadimarsky) [reply pandas/tests/io/test_sql.py:2108]: I think that's already done in the first test here
https://github.com/pandas-dev/pandas/blob/dbf0cfa9337789a24afbd1034bee1fb69a058b57/pandas/tests/io/test_sql.py#L2076-L2079
which calls a method that passes the engine param (in this case, an invalid one named `bad_engine`) to the function
https://github.com/pandas-dev/pandas/blob/dbf0cfa9337789a24afbd1034bee1fb69a058b57/pandas/tests/io/test_sql.py#L580-L595

Does this suffice? Or did you have something else in mind?

The test that I left as `# TODO` is for ensuring that when importing 3rd party libraries for the upcoming engines it raises the correct Exceptions if it can't import them

File: pandas/io/sql.py
Comment (yehoshuadimarsky) [reply pandas/io/sql.py]: fixed

**Review by jreback — APPROVED** [review]
lgtm ping on green

File: pandas/tests/io/test_sql.py
Comment (yehoshuadimarsky) [reply pandas/tests/io/test_sql.py:2108]: @jreback you marked this as resolved, assuming it's ok then?

**Comment by yehoshuadimarsky** [comment]
> lgtm ping on green

ok will do, thanks!

**Comment by yehoshuadimarsky** [comment]
The same random CI error for the nan/inf tests in MySQL... anyone know how to resolve? Pretty sure it has nothing to do with my changes

**Comment by jreback** [comment]
> The same random CI error for the nan/inf tests in MySQL... anyone know how to resolve? Pretty sure it has nothing to do with my changes

this *should* be xfailed already, you have merged master?

**Comment by yehoshuadimarsky** [comment]
> > The same random CI error for the nan/inf tests in MySQL... anyone know how to resolve? Pretty sure it has nothing to do with my changes
> 
> this _should_ be xfailed already, you have merged master?

Pretty sure I did. 

This is the offending test, permalinked from my PR branch:

https://github.com/yehoshuadimarsky/pandas/blob/c34c97b7f70297223ae48cc7b3893a824eb273e6/pandas/tests/io/test_sql.py#L2029-L2051

And here's the permalink from my branch of the code that raises the error:

https://github.com/yehoshuadimarsky/pandas/blob/c34c97b7f70297223ae48cc7b3893a824eb273e6/pandas/io/sql.py#L1342-L1351

Although in fairness, I moved that code out of `to_sql` and into the `SQLAlchemyEngine` class

File: pandas/io/sql.py
Comment (xhochy) [inline pandas/io/sql.py:1381]: Here you could also use the entrypoint mechanism to do the actual engine load. 

```suggestion
        return SQLAlchemyEngine()
    else:
        return {e.name: e for e in pkg_resources.iter_entry_points('pandas_sql_engine')}[engine].load()()
```

**Review by xhochy — COMMENTED** [review]
This looks good in general. I've added some suggestions how you could load engines via entrypoints. I would like to avoid that we only can add engine via PRs to `pandas`. This would make `pandas`' codebase just much much larger and would also slow down the fixes to the engines as we would need reviewers from the engine library and pandas-core to merge it.

**Comment by lithomas1** [comment]
@yehoshuadimarsky I think CI is because ```-np.inf``` raises a different error than ```np.inf```. Which I think is ```(1054, "Unknown column 'infe0' in 'field list'")```. 
This then is not caught by 
https://github.com/pandas-dev/pandas/blob/c34c97b7f70297223ae48cc7b3893a824eb273e6/pandas/io/sql.py#L1345-L1351
when it should be. 
This was caught previously by 
https://github.com/pandas-dev/pandas/blob/49c779118c2d473c60565d9010e119bb62d06f74/pandas/io/sql.py#L1595-L1599
(code from master removed in this PR, which I think is a little too broad.)

Can you try adding catching ```(1054, "Unknown column 'infe0' in 'field list'")``` too?

File: pandas/core/config_init.py
Comment (yehoshuadimarsky) [reply pandas/core/config_init.py:669]: I don't know... I've never used `entrypoint`s before, makes me a little uneasy to use them without understanding them, I'd need to learn some more about it... But we don't do this for any other I/O engines like Parquet or Excel, so why not keep the same here?

@jreback what do you think?

**Comment by yehoshuadimarsky** [comment]
> @yehoshuadimarsky I think CI is because `-np.inf` raises a different error than `np.inf` in PyMySQL. Which I think is `(1054, "Unknown column 'infe0' in 'field list'")`.
> This then is not caught by
> https://github.com/pandas-dev/pandas/blob/c34c97b7f70297223ae48cc7b3893a824eb273e6/pandas/io/sql.py#L1345-L1351
> 
> 
> when it should be.
> This was caught previously by
> https://github.com/pandas-dev/pandas/blob/49c779118c2d473c60565d9010e119bb62d06f74/pandas/io/sql.py#L1595-L1599
> 
> 
> (code from master removed in this PR, this code seems a little too broad to me, and probably needs to be narrowed down.)
> Can you try catching `(1054, "Unknown column 'infe0' in 'field list'")` too?

Thanks!! So because it's a regex, can I just change it to `msg = "(1054, \"Unknown column 'inf(e0)?' in 'field list'\")"` (adding the `(e0)?` as an optional part)? I tested it out here https://regex101.com/r/cBUuvd/1

**Comment by yehoshuadimarsky** [comment]
Nope CI fix didn't work 😢

**Comment by jreback** [comment]
IIRC we had an xfail in place for this partcular test. go ahead an dput it back and can be tackled there.

**Comment by lithomas1** [comment]
@jreback This is not a flaky test. This is failing because we are not raising the correct error message. See #34431. Xfailed case is also xfailed here I think and is not the problem. 
@yehoshuadimarsky We're a little closer now with 2 tests failing instead of 4. You need to also catch ```ProgrammingError: inf can not be used with MySQL```. Why? See https://github.com/PyMySQL/mysqlclient/issues/246, where they actually catch the inf case and raise a nice error. That appears in our CI with the latest pymysql, but with old pymysql(e.g. 0.8.1 where they don't raise a nice error), you need to catch original. (```OperationalError: (1054, "Unknown column ... ```)

**Comment by yehoshuadimarsky** [comment]
> @jreback This is not a flaky test. This is failing because we are not raising the correct error message. See #34431. Xfailed case is also xfailed here I think and is not the problem.
> @yehoshuadimarsky We're a little closer now with 2 tests failing instead of 4. :) You need to also catch `ProgrammingError: inf can not be used with MySQL`. Why? See [PyMySQL/mysqlclient#246](https://github.com/PyMySQL/mysqlclient/issues/246), where they actually catch the inf case and raise a nice error. That appears in our CI with the latest pymysql, but with old pymysql(e.g. 0.8.1 where they don't raise a nice error), you need to catch original. (`OperationalError: (1054, "Unknown column ... `)

Great. So I catch these 2?
- `(1054, "Unknown column 'inf' in 'field list'")`
- `(1054, "Unknown column 'infe0' in 'field list'")`

**Comment by lithomas1** [comment]
Also programmmingerror.

File: pandas/core/config_init.py
Comment (xhochy) [reply pandas/core/config_init.py:669]: 1. I would expect the number of engines to be larger in this case here. Personally, I would write at least an engine each for `snowflake`, `turbodbc` and `postgres`. 
2. Databases are harder to test than file formats as you need to start additional services. Integrating all of them into the `pandas` CI would probably be quite heavy at the end.
3. I'm really excited about ExtensionArrays as this gives us a nice way to extend the `pandas` functionality with additions that are experimental and wouldn't be merged into core pandas. They don't need a technique like entrypoints as you instantiate them explicitly by using the class constructor of the ExtensionArray. This isn't the case for the database engines as you only specify the engine via a string.
4. We can use a different mechanism but entrypoints have the nice effect of being very light-weight and discovery works without any Python code of the additional packages to be run before any `pandas` code, i.e. you don't need to ensure that you have imported the engine code before any other code uses `pandas`.

**Comment by yehoshuadimarsky** [comment]
For the life of me can't figure this out: https://stackoverflow.com/questions/67358166

**Comment by yehoshuadimarsky** [comment]
@lithomas1 shoot I still can't get the CI to pass, any suggestions?

File: pandas/io/sql.py
Comment (lithomas1) [inline pandas/io/sql.py:1350]:
```suggestion
            if re.search(msg, err_text):
```

**Comment by yehoshuadimarsky** [comment]
@lithomas1 nope still failed CI

File: pandas/io/sql.py
Comment (lithomas1) [inline pandas/io/sql.py:1348]:
```suggestion
            )|inf can not be used with MySQL"""
```
One last try? I might have gotten the error message wrong its probably this ```sqlalchemy.exc.ProgrammingError: (pymysql.err.ProgrammingError) inf can not be used with MySQL```, not ```pymysql.err.ProgrammingError: inf can not be used with MySQL``` (as I previously thought).

**Comment by lithomas1** [comment]
@yehoshuadimarsky Yay, looks like the CI fix worked finally. I think the last thing to do is to evaluate @xhochy 's proposal. I'm not a core team member so I'll leave this to @jreback and @charlesdong1991.

**Comment by yehoshuadimarsky** [comment]
> @yehoshuadimarsky Yay, looks like the CI fix worked finally. I think the last thing to do is to evaluate @xhochy 's proposal. I'm not a core team member so I'll leave this to @jreback and @charlesdong1991.

Thanks for all your help!

File: pandas/core/config_init.py
Comment (jreback) [reply pandas/core/config_init.py:669]: I think we could certainly add this (we already use entry points for plotting), but let's do as a followup (pls create an issue)

**Review by jreback — APPROVED** [review]
thanks @yehoshuadimarsky very nice (comment for followup)

**Comment by jreback** [comment]
thanks @yehoshuadimarsky very nice!

**Comment by jreback** [comment]
@yehoshuadimarsky can you do a followup to add a whatsnew note / update io.rst if needed anywhere.

**Comment by yehoshuadimarsky** [comment]
> @yehoshuadimarsky can you do a followup to add a whatsnew note / update io.rst if needed anywhere.

Yes, just for this change? There's not much here that changed to the external user, it's just refactoring to support more future engines.

**Comment by jreback** [comment]
yeah just a quick one for the addition of the option (even though it doesn't do anything different yet)

**Pull request merged**
